walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,92 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Arrays;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.HashSet;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> attendees = request.getAttendees();
+    long duration = request.getDuration();
+
+    //Case in which the requested meeting has no attendies
+    if(attendees.isEmpty()){
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    //Case in which the requested meeting has a duration that is too long/short
+    if(duration > TimeRange.WHOLE_DAY.duration() || duration < 0){
+        return Arrays.asList();
+    } 
+
+    //List of time ranges of other events that attendees are attending
+    List<TimeRange> timeRanges = new ArrayList<TimeRange>();
+    TimeRange currentRange;
+    for(String person : attendees){
+        for(Event event: events){
+            currentRange = event.getWhen();
+            if(event.getAttendees().contains(person) && !timeRanges.contains(currentRange)){
+                timeRanges.add(currentRange);
+            }
+        }
+    }
+
+    //Case in which no events on calendar so no conflicts
+    if(timeRanges.isEmpty()){
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    //Sort timeRanges by starting time 
+    Collections.sort(timeRanges, TimeRange.ORDER_BY_START);
+
+    Collection<TimeRange> availTimes = new ArrayList<TimeRange>(); //Time ranges that are available for the meeting request
+    int startOfAvail = TimeRange.START_OF_DAY; //The start of an available time range
+    int endOfAvail; 
+    int timeRangesCtr = 0; //Index of current position in array list "timeRanges"
+    TimeRange lastTimeRange = null; //Last time range that has been accounted for from timeRanges
+    TimeRange nextTimeRange = timeRanges.get(0);; //Next time range that needs to be accounted for
+
+    if(nextTimeRange.start() == TimeRange.START_OF_DAY){
+        //If earliest event starts at 00:00 then update startOfAvail 
+        //to be the end time of the first event
+        startOfAvail = nextTimeRange.end();
+
+        //First event is accounted for, so update lastTimeRange
+        lastTimeRange = nextTimeRange;
+        timeRangesCtr++;
+    } 
+
+    while(startOfAvail < TimeRange.END_OF_DAY && timeRangesCtr < timeRanges.size()){ 
```

Comment:
You could add a dummy first event (0,0) to handle some corner cases. Also you could add a dummy last event (24 * 60, 24 * 60) as the last event.

Reply:
Would you be able to tell me more about this idea tomorrow during our sync please?